core/src/util.rs

```diff
@@ -27,3 +27,187 @@ impl<T: Seq> SeqGenerator<T> {
         mem::replace(&mut self.0, value)
     }
 }
+
+pub trait PacketData: Copy {
+    fn write(self, vec: &mut Vec<u8>);
+
+    fn size_hint(&self) -> usize;
+}
+
+pub trait IntoPacketData<T = Self> {
+    type Data: PacketData;
+
+    fn convert(data: T) -> Self::Data;
+}
+
+// `()` represents an empty packet
+impl PacketData for () {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        0
+    }
+
+    #[inline]
+    fn write(self, _: &mut Vec<u8>) {}
+}
+
+// `(A, B)` represents the A and B concatenated
+impl<A, B> PacketData for (A, B)
+where
+    A: PacketData,
+    B: PacketData,
+{
+    #[inline]
+    fn size_hint(&self) -> usize {
+        self.0.size_hint() + self.1.size_hint()
+    }
+
+    #[inline]
+    fn write(self, vec: &mut Vec<u8>) {
+        self.0.write(vec);
+        self.1.write(vec);
+    }
+}
+
+macro_rules! impl_packet_data_num {
+    ( $($t:ty : $n:literal), *) => {
+        $(
+            impl PacketData for $t {
+                #[inline]
+                fn size_hint(&self) -> usize {
+                    $n
```

Comment:
Using [size_of](https://doc.rust-lang.org/std/mem/fn.size_of.html) is probably good enough here.

Reply:
Yes. My first solution needed the type `[u8; $n]` explicitly, but that's not the case anymore. I'll update it.